### PR TITLE
FailsAt annotation 

### LINKED ISSUE #549: Provide a @FailAt annotation
Inspired by https://twitter.com/mikeveerman/status/1464162014444638258

It would be nice to provide an annotation that fails at a specified date, with on optional reason. Something like this:

```java
@FailAt(date = "2022-01-01", reason = "An optional reason")
@Test
void should_fail_at_the_provided_date() {
    // ...
}
```

This could be a great way to remind oneself of unpaid technical debt but could probably also be used in other contexts.

### PR DESCRIPTION
**Pull request by Bukama**
Proposed commit message:

```
FailAt Extension (#549 / #814) 

The FailAt Extension provides to fail annotated test methods or
classes when a given date is reached.

closes #549
PR: #814
```

### CODE REVIEW

File: src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+		}
+
+		@Test
+		@FailAt(reason = "Everything was better in the past!", date = "1993-01-01")
+		void testIsAnnotatedWithDateInThePast() {
+
+		}
+
+		// How can we test today? This way gives "attribute must be constant compile error"
```
Comment (Bukama) [inline src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java:93]: @Michael1993  Do you have an idea? We don't test today in the `@DisabledUntil` extension either so we miss this test edge case there too.

**Comment by Bukama** [comment]
Aside from my question about how to test "today" the PR would be no draft. So feel free to review the other changes too.

File: src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java (hunk 1 above)
Comment (Michael1993) [reply src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java:93]: We can create a mock `ExecutionContext` and call the extension directly. Not sure if all the orchestration is worth it, but that is one way to do it...

Comment (Bukama) [reply src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java]: > We can create a mock `ExecutionContext` and call the extension directly. Not sure if all the orchestration is worth it, but that is one way to do it...

Decided to trust the JDK guys that they know how to implement `before` and that you will see if my if condition is correct by looking at it during PR review.

File: src/demo/java/org/junitpioneer/jupiter/FailAtExtensionDemo.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	void test() {
+		// Test will fail as soon as 1st of January 2025 is reached.
+	}
+	// end::fail_at_simple[]
+
+	// tag::fail_at_with_reason[]
+	@Test
+	@DisabledUntil(reason = "We are not allowed to call that method anymore", date = "2025-01-01")
+	void testWithReason() {
+		// Test will fail as soon as 1st of January 2025 is reached.
```
Comment (beatngu13) [inline src/demo/java/org/junitpioneer/jupiter/FailAtExtensionDemo.java:30]:
```suggestion
		// Test will fail with the given reason as soon as 1st of January 2025 is reached.
```

File: docs/fail-at.adoc [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+Invalid or unparsable date strings lead to an `ExtensionConfigurationException`.
+
+The `@FailAt annotation may optionally be declared with a reason to document why the annotated test class or test method fails as soon as the date is reached:
+
+[source,java,indent=0]
+----
+include::{demo}[tag=fail_at_with_reason]
+----
+
+The `@FailAt` annotation can be used on the class and method level, it will be inherited from higher level containers:
```
Comment (beatngu13) [inline docs/fail-at.adoc:39]:
```suggestion
The `@FailAt` annotation can be used on the class and method level, it will be inherited from higher-level containers:
```

File: src/main/java/org/junitpioneer/jupiter/FailAt.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * <p>It may optionally be declared with a reason to document why the annotated test class or test
+ * method should fail at the given date.</p>
+ *
+ * <p>{@code @FailAt} can be used on the method and class level. It can only be used once per method or class,
+ * but is inherited from higher-level containers.</p>
+ *
+ * <p><strong>WARNING:</strong> This annotation allows the user to move an assumption out of one or multiple test
+ * method's code into an annotation. But this comes at a cost - applying {@code @FailAt} can make the test suite
+ * non-reproducible. If a passing test is run again after the "failAt" date that build would fail. A report entry is
+ * issued for every test that does not fail until a certain date.</p>
```
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/FailAt.java:34]:
```suggestion
 * method's code into an annotation. But this comes at a cost: applying {@code @FailAt} can make the test suite
 * non-reproducible. If a passing test is run again after the specified date, that build would fail. A report entry is
 * issued for every test that does not fail until a certain date.</p>
```

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		return findClosestEnclosingAnnotation(context, FailAt.class).map(FailAt::date).map(this::parseDate);
+	}
+
+	private LocalDate parseDate(String dateString) {
+		try {
+			return LocalDate.parse(dateString, ISO_8601);
+		}
+		catch (DateTimeParseException ex) {
+			throw new ExtensionConfigurationException(
+				"The `failAtDate` string '" + dateString + "' is no valid ISO-8601 string.", ex);
```
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:53]:
```suggestion
				"The `failAtDate` string '" + dateString + "' is not a valid ISO-8601 string.", ex);
```

File: src/main/java/org/junitpioneer/jupiter/FailAt.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+	/**
+	 * The reason this annotated test class or test method should fail as soon as the given date is reached.
+	 */
+	String reason() default "";
+
+	/**
+	 * The date from which this annotated test class or test method should fail as an ISO 8601 string in the
+	 * format yyyy-MM-dd, e.g. 2023-05-28.
+	 * The test will be executed regularly if that date is not yet reached.
```
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/FailAt.java:52]:
```suggestion
	 * format yyyy-MM-dd, e.g. 2023-05-28. The test will be executed regularly if that date is not yet reached.
```

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ */
+class FailAtExtension implements ExecutionCondition {
+
+	private static final DateTimeFormatter ISO_8601 = DateTimeFormatter.ISO_DATE;
+
+	@Override
+	public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
+		return getFailAtDateFromAnnotation(context)
+				.map(failAtDate -> evaluateFailAtDate(context, failAtDate))
+				.orElse(enabled("No @FailAt annotation found on element"));
```
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:40]:
```suggestion
				.orElse(enabled("No @FailAt annotation found on element."));
```

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java (hunk 5 above)
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:53]:
```suggestion
				"The `failAtDate` string '" + dateString + "' is no valid ISO 8601 string.", ex);
```

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		} else {
+			String reportEntry = format(
+				"The current date %s is after or on the `date` %s, so `@FailAt` fails the test \"%s\". Please remove the annotation.",
+				failAtDate.format(ISO_8601), today.format(ISO_8601), context.getUniqueId());
+			context.publishReportEntry(FailAtExtension.class.getSimpleName(), reportEntry);
+
+			String message = format("The current date %s is after or on the `date` %s", today.format(ISO_8601),
+				failAtDate.format(ISO_8601));
+
+			throw new ExtensionConfigurationException(message);
```
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:76]:
```suggestion
			throw new AssertionFailedError(message, failAtDate, today);
```

Or just the message - up to you.

I don't think ExtensionConfigurationException is the right exception to throw here.

**Review by Michael1993 — CHANGES_REQUESTED** [review]

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	}
+
+	private ConditionEvaluationResult evaluateFailAtDate(ExtensionContext context, LocalDate failAtDate) {
+		LocalDate today = LocalDate.now();
+		boolean isBefore = today.isBefore(failAtDate);
+
+		if (isBefore) {
+			String reportEntry = format(
+				"The `date` %s is after the current date %s, so `@FailAt` did not fails the test \"%s\". It will do so when the date is reached.",
```
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:63]:
```suggestion
				"The `date` %s is after the current date %s, so `@FailAt` did not fail the test \"%s\". It will do so when the date is reached.",
```

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+				failAtDate.format(ISO_8601), today.format(ISO_8601), context.getUniqueId());
+			context.publishReportEntry("FailAt", reportEntry);
+			return enabled(reportEntry);
+		} else {
+			String reportEntry = format(
+				"The current date %s is after or on the `date` %s, so `@FailAt` fails the test \"%s\". Please remove the annotation.",
+				failAtDate.format(ISO_8601), today.format(ISO_8601), context.getUniqueId());
```
Comment (beatngu13) [inline src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:70]: You are formatting `failAtDate` and `today` here and below. Actually, you are formatting both variables also in the `isBefore` case. Maybe extract the formatted dates outside the if block?

File: src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+	@DisplayName("Should not fail a test with `date` in the future")
+	void shouldNotFailTestWithFailAtDateInTheFuture() {
+		final ExecutionResults results = PioneerTestKit
+				.executeTestMethod(FailAtTestCases.class, "testIsAnnotatedWithDateInTheFuture");
+		assertThat(results).hasSingleSucceededTest();
+		assertThat(results).hasSingleReportEntry().firstValue().contains("2199-01-01", "did not fails the test");
+	}
+
+	@Test
+	@DisplayName("Should fail nested test with `date` in the past when meta annotated by higher level container")
```
Comment (beatngu13) [inline src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java:65]:
```suggestion
	@DisplayName("Should fail nested test with `date` in the past when meta annotated by higher-level container")
```

File: src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+		assertThat(results).hasNoReportEntries();
+	}
+
+	@Test
+	@DisplayName("Should enable test with unparsable `date`` string")
+	void shouldEnableTestWithUnparsableUntilDateString() {
+		final ExecutionResults results = PioneerTestKit
+				.executeTestMethod(FailAtTestCases.class, "testUnparsableUntilDateString");
+		assertThat(results).hasSingleStartedTest();
+		assertThat(results).hasSingleFailedTest();
```
Comment (beatngu13) [inline src/test/java/org/junitpioneer/jupiter/FailAtExtensionTests.java:40]: In case of the failed test (also below), I'd assert the thrown exception and also its message.

**Review by beatngu13 — CHANGES_REQUESTED** [review]

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java (hunk 5 above)
Comment (Bukama) [reply src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:53]: I think both are valid, but will change it :)

File: src/main/java/org/junitpioneer/jupiter/FailAtExtension.java (hunk 8 above)
Comment (Bukama) [reply src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:76]: > Or just the message - up to you.
> 
> I don't think ExtensionConfigurationException is the right exception to throw here.

`AssertionFailedError` is an exception for tests by `opentest4j` (see [javadoc](https://ota4j-team.github.io/opentest4j/docs/1.0.0/api/org/opentest4j/AssertionFailedError.html) not for "production" code, so this is not usable here (couldn't be imported anyway as the dependency is has test scope).

I'm open to whatever exception we use, but I think it's some kind of configuration error, when seeing the `failAt` date is reached and therefore the extension / annotation should not be used anymore with the current configuration. I / we also use the `ExtensionConfigurationException` in other situation of this extension and the `DisabledUntil` where the date is not properly set up.

@beatngu13 What do you think?

Comment (beatngu13) [reply src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:76]: JUniter Jupiter (API) pulls in OpenTest4J as a [compile dependency](https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter-api/5.10.1), so do we? Since we are testing library, that is just fine. We already use several exception types from there:

https://github.com/search?q=repo%3Ajunit-pioneer%2Fjunit-pioneer+opentest4j+language%3AJava+path%3A%2F%5Esrc%5C%2Fmain%5C%2Fjava%5C%2Forg%5C%2Fjunitpioneer%5C%2F%2F&type=code

Comment (Bukama) [reply src/main/java/org/junitpioneer/jupiter/FailAtExtension.java:76]: Hm. Okay, will use it.

**Comment by Bukama** [comment]
Updates with your review results. If I have not overseen anything, I think that PR is done and ready to merge.

**Review by beatngu13 — APPROVED** [review]
